src/modules/files/stat.rs

```
@@ -0,0 +1,150 @@
+            },
+
+            _ => { return Err(handle.response.not_supported(request)); }
+
+        }
+
+    }
+
+}
+
+#[derive(Serialize)]
+struct StatResult {
+    pub exists: bool,
+    pub is_dir: Option<bool>,
+    pub mode: Option<String>,
+    pub size: Option<i64>,
+    pub owner: Option<String>,
+    pub group: Option<String>,
+}
+
+const NON_EXISTING_RESULT: StatResult = StatResult{
+    exists: false,
+    is_dir: None,
+    mode: None,
+    size: None,
+    owner: None,
+    group: None,
+};
+
+fn stat_file(handle: &Arc<TaskHandle>, request: &Arc<TaskRequest>, path: &String) -> Result<StatResult, Arc<TaskResponse>> {
+    let mode_option = handle.remote.get_mode(request, path)?;
+    match mode_option {
+        Some(mode) => {
+            let is_dir = handle.remote.get_is_directory(request, path)?;
+            let ownership_option = handle.remote.get_ownership(request, path)?;
+            match (is_dir, ownership_option) {
+                (true, Some((owner, group))) => Ok(StatResult{
+                    exists: true,
+                    is_dir: Some(true),
+                    mode: Some(mode),
+                    size: None,
+                    owner: Some(owner),
+                    group: Some(group),
+                }),
+                (false, Some((owner, group))) => {
+                    let size_option = handle.remote.get_size(request, path)?;
+                    match size_option {
+                        Some(size) => Ok(StatResult{
+                            exists: true,
+                            is_dir: Some(false),
+                            mode: Some(mode),
+                            size: Some(size),
+                            owner: Some(owner),
+                            group: Some(group),
+                        }),
+                        None => Ok(NON_EXISTING_RESULT),
+                    }
+                },
+                _ => Ok(NON_EXISTING_RESULT),
```

we're allowing the mode to fail rather than requiring every single step to decide whether the file now exists or not, so I think this part can be removed, just let the error bubble up if it happens in the middle of the module call.  It's arguable but I think its' better to know something weird happened in the module run anyway, than it trying to reflect a changing state during execution.... ie.. let it fail rather than returning that it does't exist.

---

Fail in what way? As an error or as a panic?

---

The command modules will fail all the way up to Arc<TaskResponse> when they have non-zero return codes if they are executed as Checked

---

I mean like, do `Err(...)` instead of `Ok(...)` so it's just an error, or on the line with `handle.remote.get_ownership(...)?` should I add an `.unwrap()` at the end and treat it like a panic?

---

I feel like I'm making you change too many things and am willing to tweak this one if I'm bad at explaining.  It's like swapping a couple of lines around and I'm happy to do that to make you not have to work on this repeatedly :)

On panics, no, the code should always return Err(...) and never panic inside of modules.  A panic in one thread will kill all the other threads.

---

Then yes please. Commit to this PR or just merge and fix later.

I'm quite drained from fixing this PR.

---

Yeah that is what I was trying to warn people about in the contrib guide :)  Appreciate the work and can change remaining easy bits.